giraph-core/src/main/java/org/apache/giraph/writable/kryo/GiraphClassResolver.java

```diff
@@ -0,0 +1,333 @@
+  private static ZooKeeperExt ZK;
+  /** Zookeeper path for automatic class registrations */
+  private static String KRYO_REGISTERED_CLASS_PATH;
+  /** Minimum class ID assigned by zookeeper sequencing */
+  private static int MIN_CLASS_ID = -1;
+
+  /** Memoized class id*/
+  private int memoizedClassId = -1;
+  /** Memoized class registration */
+  private Registration memoizedClassIdValue;
+
+  /**
+   * Sets zookeeper informaton.
+   * @param zookeeperExt ZookeeperExt
+   * @param kryoClassPath Zookeeper directory path where class Name-ID
+   *                      mapping is stored.
+   */
+  public static void setZookeeperInfo(ZooKeeperExt zookeeperExt,
+                                      String kryoClassPath) {
+    ZK = zookeeperExt;
+    KRYO_REGISTERED_CLASS_PATH = kryoClassPath;
+  }
+
+  /**
+   * Creates a new node for the given class name.
+   * Creation mode is persistent sequential, i.e.
+   * ZK will always create a new node . There could be
+   * multiple entries for the same class name but since
+   * the lowest index is used, this is not a problem.
+   * @param className Class name
+   */
+  public static void createClassName(String className) {
+    try {
+      String path = KRYO_REGISTERED_CLASS_PATH + "/" + className;
+      ZK.createExt(path,
+              null,
+              ZooDefs.Ids.OPEN_ACL_UNSAFE,
+              CreateMode.PERSISTENT_SEQUENTIAL,
+              true);
+    } catch (KeeperException e) {
+      throw new IllegalStateException(
+              "Failed to create class " + className, e);
+    } catch (InterruptedException e) {
+      throw new IllegalStateException(
+              "Interrupted while creating " + className, e);
+    }
+  }
+
+  /**
+   * Refreshes class-ID mapping from zookeeper.
+   * Not thread safe.
+   */
+  public static void refreshCache() {
+    try {
+      ZK.createOnceExt(KRYO_REGISTERED_CLASS_PATH,
+              null,
+              ZooDefs.Ids.OPEN_ACL_UNSAFE,
+              CreateMode.PERSISTENT,
+              true);
+    } catch (KeeperException e) {
+      throw new IllegalStateException(
+              "Failed to refresh kryo cache " +
+                      KRYO_REGISTERED_CLASS_PATH, e);
+    } catch (InterruptedException e) {
+      throw new IllegalStateException(
+              "Interrupted while refreshing kryo cache " +
+                      KRYO_REGISTERED_CLASS_PATH, e);
+    }
+
+    List<String> registeredList;
+    try {
+      registeredList =
+              ZK.getChildrenExt(KRYO_REGISTERED_CLASS_PATH,
+                      false,
+                      true,
+                      false);
+    } catch (KeeperException e) {
+      throw new IllegalStateException(
+        "Failed to retrieve child nodes for " + KRYO_REGISTERED_CLASS_PATH, e);
+    } catch (InterruptedException e) {
+      throw new IllegalStateException(
+        "Interrupted while retrieving child nodes for " +
+                KRYO_REGISTERED_CLASS_PATH, e);
+    }
+
```

Comment:
Can we add some debug logging here that prints the returned list? With the usual pattern:

if (LOG.isDebugEnabled()) {
      LOG.debug("...");
}

Reply:
Ok.